store/bigtable/bigtable.go

```diff
@@ -0,0 +1,471 @@
+			}
+			for _, ttl := range ttls {
+				policy, ok := existingFamilies[formatFamily(ttl)]
+				if !ok {
+					log.Infof("btStore: column family %s/%s does not exist. creating it", cfg.TableName, formatFamily(ttl))
+					err = adminClient.CreateColumnFamily(ctx, cfg.TableName, formatFamily(ttl))
+					if err != nil {
+						return nil, fmt.Errorf("btStore: failed to create cf %s/%s. %s", cfg.TableName, formatFamily(ttl), err)
+					}
+					err = adminClient.SetGCPolicy(ctx, cfg.TableName, formatFamily(ttl), bigtable.MaxAgePolicy(time.Duration(ttl/(60*60))*time.Hour))
+					if err != nil {
+						return nil, fmt.Errorf("btStore: failed to set GCPolicy of %s/%s. %s", cfg.TableName, formatFamily(ttl), err)
+					}
+				} else if policy == "" {
+					log.Infof("btStore: column family %s/%s exists but has no GCPolicy. creating it", cfg.TableName, formatFamily(ttl))
+					err = adminClient.SetGCPolicy(ctx, cfg.TableName, formatFamily(ttl), bigtable.MaxAgePolicy(time.Duration(ttl/(60*60))*time.Hour))
+					if err != nil {
+						return nil, fmt.Errorf("btStore: failed to set GCPolicy of %s/%s. %s", cfg.TableName, formatFamily(ttl), err)
+					}
+				}
+			}
+		}
+	}
+
+	client, err := bigtable.NewClient(ctx, cfg.GcpProject, cfg.BigtableInstance)
+	if err != nil {
+		return nil, fmt.Errorf("btStore: failed to create bigtable client. %s", err)
+	}
+
+	s := &Store{
+		client:           client,
+		tbl:              client.Open(cfg.TableName),
+		shutdown:         make(chan struct{}),
+		writeQueues:      make([]chan *mdata.ChunkWriteRequest, cfg.WriteConcurrency),
+		writeQueueMeters: make([]*stats.Range32, cfg.WriteConcurrency),
+		readQueue:        make(chan struct{}, cfg.ReadConcurrency),
+		cfg:              cfg,
+	}
+	for i := 0; i < cfg.WriteConcurrency; i++ {
+		s.wg.Add(1)
+		// The writeQueue is comprised of 2 components.  The chan and a buffer in the procesWriteQueue thread.
+		// The buffer in the thread has a limit of writeMaxFlushSize.
+		s.writeQueues[i] = make(chan *mdata.ChunkWriteRequest, cfg.WriteQueueSize-cfg.WriteMaxFlushSize)
+		s.writeQueueMeters[i] = stats.NewRange32(fmt.Sprintf("store.bigtable.write_queue.%d.items", i+1))
+		go s.processWriteQueue(s.writeQueues[i], s.writeQueueMeters[i])
+	}
+
+	return s, nil
+}
+
+func (s *Store) Stop() {
+	close(s.shutdown)
+	s.wg.Wait()
+	err := s.client.Close()
+	if err != nil {
+		log.Errorf("Error closing bigtable client. %s", err)
+	}
+}
+
+func (s *Store) Add(cwr *mdata.ChunkWriteRequest) {
+	sum := int(cwr.Key.MKey.Org)
+	for _, b := range cwr.Key.MKey.Key {
+		sum += int(b)
+	}
+	which := sum % len(s.writeQueues)
+	s.writeQueueMeters[which].Value(len(s.writeQueues[which]))
+	s.writeQueues[which] <- cwr
+}
+
+func (s *Store) processWriteQueue(queue chan *mdata.ChunkWriteRequest, meter *stats.Range32) {
+	defer s.wg.Done()
+	// monitor the queue length.  We use a separate goroutine so that monitoring will still
+	// occur even if reading from the queue blocks.
```

Comment:
if a queue spikes up and drains within a collection interval, the meter will have no idea and monitoring is off.
as it's a fairly lightweight operation, I think it's better to update the meter upon all relevant queue channel operations

Reply:
so I think we need the combination of both approaches.
at-chan-operation monitoring to catch quick changes, and side-goroutine monitoring to consistently get a value when the queue is blocked. (because the meter resets when it is reported)
there's a bunch of places throughout the code where we should do this, I think

Reply:
As writes are flushed at set intervals, we make 1000's of reads/second on from the queue.  At high frequency, even lightweight operations become expensive.   With the periodic flush model, we already experience a burst of high CPU usage.  We should be working to reduce that, not make it worse.

Finally, the entire purpose of monitoring the queue is to ensure write threads are able to keep up with the volume.  If the queue fills up and drains between our 1second polls of the queue length, then the threads are working fast enough.

Reply:
sounds good